common/scala/src/main/scala/whisk/core/connector/MessageConsumer.scala

```diff
@@ -38,15 +53,188 @@ trait MessageConsumer {
      * from the connector.
      */
-    def commit()
-
-    /**
-     * Calls process for every message received. Process receives a tuple
-     * (topic, partition, offset, and message as byte array).
-     */
-    def onMessage(process: (String, Int, Long, Array[Byte]) => Unit): Unit
+    def commit(): Unit
 
     /** Closes consumer. */
     def close(): Unit
 
 }
+
+object MessageFeed {
+    protected sealed trait FeedState
+    protected[connector] case object Uninitialized extends FeedState
+    protected[connector] case object FillingPipeline extends FeedState
+    protected[connector] case object DrainingPipeline extends FeedState
+
+    protected sealed trait FeedData
+    private case object NoData extends FeedData
+
+    /** Indicates the consumer is ready to accept messages from the message bus for processing. */
+    object Ready
+
+    /** Steady state message, indicates capacity in downstream process to receive more messages. */
+    object Processed
+
+    /** Indicates the fill operation has completed. */
+    private object FillCompleted
+}
+
+/**
+ * This actor polls the message bus for new messages and dispatches them to the given
+ * handler. The actor tracks the number of messages dispatched and will not dispatch new
+ * messages until some number of them are acknowledged.
+ *
+ * This is used by the invoker to pull messages from the message bus and apply back pressure
+ * when the invoker does not have resources to complete processing messages (i.e., no containers
+ * are available to run new actions). It is also used in the load balancer to consume active
+ * ack messages.
+ * When the invoker releases resources (by reclaiming containers) it will send a message
+ * to this actor which will then attempt to fill the pipeline with new messages.
+ *
+ * The actor tries to fill the pipeline with additional messages while the number
+ * of outstanding requests is below the pipeline fill threshold.
+ */
+@throws[IllegalArgumentException]
+class MessageFeed(
+    description: String,
+    logging: Logging,
+    consumer: MessageConsumer,
+    maximumHandlerCapacity: Int,
+    longPollDuration: FiniteDuration,
+    handler: Array[Byte] => Future[Unit],
+    autoStart: Boolean = true,
+    logHandoff: Boolean = true)
+    extends FSM[MessageFeed.FeedState, MessageFeed.FeedData] {
+    import MessageFeed._
+
+    // double-buffer to make up for message bus read overhead
+    val maxPipelineDepth = maximumHandlerCapacity * 2
+    private val pipelineFillThreshold = maxPipelineDepth - consumer.maxPeek
+
+    require(consumer.maxPeek <= maxPipelineDepth, "consumer may not yield more messages per peek than permitted by max depth")
+
+    private val outstandingMessageQueue = new ConcurrentLinkedQueue[(String, Int, Long, Array[Byte])]()
+    private val outstandingMessageCount = new AtomicInteger() // avoid non-contant time size calculation on queue
+    private val handlerCapacity = new AtomicInteger(maximumHandlerCapacity)
```

Comment:
You shouldn't need concurrent datastructures here, if you don't close over inernal state in `Futures` inside of the Actor. You can achieve this by `pipe`ing the `Future` to `self` to receive the messages and handle them then. That also drops the need for the `FillComplete` message.